### PR TITLE
Add support for resource and collection literal parsing with comprehensive logics

### PR DESCRIPTION
**Pull request by xuzhg**
…nsive tests

This commit implements full support for parsing resource literals (JSON objects) and collection literals (JSON arrays) in OData URI queries.

Core Implementation:
- Added ResourceLiteralToken for representing JSON object syntax in queries
- Added CollectionLiteralToken for representing JSON array syntax
- Implemented ResourceLiteralBinder for binding resource literal tokens to ResourceConstantNode
- Implemented CollectionLiteralBinder for binding collection literal tokens
- Added ResourceConstantNode semantic AST node for resource literals
- Added CollectionConstantNode for collection literals
- Updated QueryTokenKind and InternalQueryNodeKind enums with new node types

Binders and Parsing:
- Enhanced LiteralBinder with comprehensive type binding and conversion logic
- Added null literal binding support with proper type checking
- Implemented enum literal binding from string and integer values
- Added integral type conversion (Int32 to Byte/SByte/Int16)
- Support for nested resource and collection literals
- Handle typed and untyped properties in resource literals

Documentation:
- Added comprehensive XML documentation to ResourceLiteralBinder class
- Documented all public and internal APIs
- Added detailed parameter descriptions and return value documentation

Test Coverage:
- Added ResourceLiteralBinderTests with 15 comprehensive test cases
  * Constructor validation tests
  * Type resolution scenarios (expected type, inferred type, type compatibility)
  * Nested resource literal handling
  * Collection property support
  * Error handling for incompatible types

- Added CollectionLiteralBinderTests with 18 comprehensive test cases
  * Empty collections
  * Primitive, string, and resource literal collections
  * Nested collection support
  * Root path token handling
  * Mixed type validation and error scenarios

- Enhanced LiteralBinderTests from 6 to 47 test cases
  * Basic literal types (int, string, bool, guid, etc.)
  * Null literal handling
  * Numeric type conversions
  * Enum binding tests
  * Spatial type support
  * Type conversion and promotion tests
  * Edge cases and error scenarios

- Added ResourceConstantNodeTests with 24 comprehensive test cases
  * Constructor behavior
  * Properties collection manipulation
  * Nested resource nodes
  * LiteralText handling
  * Visitor pattern implementation
  * Complex scenarios with mixed types

All tests pass successfully and build completes without errors.

### Issues

*This pull request fixes #xxx.*

### Description

*Briefly describe the changes of this pull request.*

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### Repository notes

Team members can start a CI build by adding a comment with the text `/AzurePipelines run` to a PR. A bot may respond indicating that there is no pipeline associated with the pull request. This can be ignored if the build is triggered.

Team members should **not** trigger a build this way for pull requests coming from forked repositories. They should instead trigger the build manually by setting the "branch" to `refs/pull/{prId}/merge` where `{prId}` is the ID of the PR.

### CODE REVIEW

File: src/Microsoft.OData.Core/UriParser/SemanticAst/CollectionConstantNode.cs [hunk 1]
```diff
@@ -8,101 +8,64 @@ namespace Microsoft.OData.UriParser
...
-            this.itemType = edmCollectionType.ElementType;
-            this.collectionTypeReference = collectionType;
-
-            foreach (object item in objectCollection)
-            {
-                this.collection.Add(new ConstantNode(item, item != null ? item.ToString() : "null", this.itemType));
-            }
-        }
+        [Obsolete("This property will be removed in the future, and the collection of constant nodes will be stored in the 'Items' property.")]
+        public IList<ConstantNode> Collection => Items.OfType<ConstantNode>().ToList();
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/SemanticAst/CollectionConstantNode.cs]: @copilot  show me an fix

File: src/Microsoft.OData.Core/PublicAPI/net10.0/PublicAPI.Shipped.txt [hunk 2]
```diff
@@ -1279,9 +1279,6 @@ Microsoft.OData.UriParser.CollectionComplexNode.CollectionComplexNode(Microsoft.
 Microsoft.OData.UriParser.CollectionComplexNode.Property.get -> Microsoft.OData.Edm.IEdmProperty
 Microsoft.OData.UriParser.CollectionComplexNode.Source.get -> Microsoft.OData.UriParser.SingleResourceNode
 Microsoft.OData.UriParser.CollectionConstantNode
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/PublicAPI/net10.0/PublicAPI.Shipped.txt:1281]: It's a major version change and allow the breaking changes. #won'tfix

File: src/Microsoft.OData.Core/UriParser/SemanticAst/CollectionConstantNode.cs [hunk 3]
```diff
@@ -8,101 +8,64 @@ namespace Microsoft.OData.UriParser
...
+        {
+            CollectionType = collectionType;
+            ItemType = collectionType?.ElementType();
+        }
 
         /// <summary>
-        /// Create a CollectionConstantNode
+        /// Gets the collection of ConstantNodes, ResourceConstantNode (such as, {...} ), CollectionConstantNode (such as, [....]).
+        /// Keep it since the ASP.NET Core OData or other OData libraries may need it, even though it's not used in the ODataLib.
+        /// This property will be removed in the future, and the collection of constant nodes will be stored in the "Value" property.
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/SemanticAst/CollectionConstantNode.cs]: #Fixed

File: src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs [hunk 4]
```diff
@@ -1245,11 +1278,44 @@ private void AdvanceThroughBalancedExpression(char startingCharacter, char endin
...
+                    if (this.Text.Span[this.textPos - 1] == '\\')
+                    {
+                        // double quote is escaped by a backslash, so skip both of them and continue.
+                        this.NextChar();
+                    }
+                    else
+                    {
+                        // end of double quoted string.
+                        break;
+                    }
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs:1317]: Fixed

File: src/Microsoft.OData.Core/UriParser/Parsers/UriParserHelper.cs [hunk 5]
```diff
@@ -44,6 +44,42 @@ internal static bool TryRemovePrefix(ReadOnlySpan<char> prefix, ref ReadOnlySpan
...
+            {
+                return false;
+            }
+
+            // double quoted string
+            if (text[0] == '"' && text[^1] == '"')
+            {
+                value = JsonSerializer.Deserialize<string>(text);
+                text = value.AsSpan();
+                return true;
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriParserHelper.cs]: #Resolved.

File: src/Microsoft.OData.Core/UriParser/SemanticAst/CollectionConstantNode.cs [hunk 6]
```diff
@@ -8,101 +8,61 @@ namespace Microsoft.OData.UriParser
...
-        public IList<ConstantNode> Collection
-        {
-            get
-            {
-                return new ReadOnlyCollection<ConstantNode>(this.collection);
-            }
-        }
+        [Obsolete("This property will be removed in the future, use the 'Items' property instead.")]
+        public IList<ConstantNode> Collection => new ReadOnlyCollection<ConstantNode>(Items.OfType<ConstantNode>().ToList());
+
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/SemanticAst/CollectionConstantNode.cs:41]: Will remove next iterator. Won't fix

File: src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs [hunk 7]
```diff
@@ -1221,7 +1254,7 @@ private static ExpressionTokenKind MakeBestGuessOnNoSuffixStr(ReadOnlySpan<char>
         private void AdvanceThroughBalancedExpression(char startingCharacter, char endingCharacter)
         {
             int currentBracketDepth = 1;
-
+            int position = this.textPos - 1;
             while (currentBracketDepth > 0)
             {
                 if (this.ch == '"')
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs:1260]: The implementation is correct since we need the original position to report the error.

File: src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs [hunk 8]
```diff
@@ -1245,11 +1278,44 @@ private void AdvanceThroughBalancedExpression(char startingCharacter, char endin
...
+            Debug.Assert(this.ch.HasValue && this.ch.Value == '"');
+
+            this.NextChar();
+            while (this.ch.HasValue)
+            {
+                if (this.ch.Value == '"')
+                {
+                    // a double quote inside the string value, must escape it with a backslash (\)
+                    if (this.Text.Span[this.textPos - 1] == '\\')
+                    {
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs:1305]: "\\" is valid double quoted string since it's a string containing a 'backslash'.
"\"" is valid double quoted string since it's a string containing a double quote.

"\\\" is invalid double quoted string but it's valid for ExpressionLexer since it doesn't care about the content within it.

"\\\"" is valid double quoted string it's a string with a backslash and a double quote. @copilot Can you review and vierfy the logic again?

File: src/Microsoft.OData.Core/UriParser/Binders/ResourceLiteralBinder.cs [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+namespace Microsoft.OData.UriParser
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Reflection;
+    using Microsoft.OData.Core;
+    using Microsoft.OData.Edm;
+    using Microsoft.OData.Metadata;
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/Binders/ResourceLiteralBinder.cs:15]: No, It's used to access the API defined within it.
